### PR TITLE
feat: different floor types

### PR DESCRIPTION
**Pull request by bahrmichael**
Closes https://github.com/bahrmichael/factorio-tycoon/issues/298

### CODE REVIEW

**Comment by winex** [comment]
hmm, `dry-dirt` for roads looks way better!
i also tried:
- `landfill` under `simple-house`
- `stone-brick` under `residential-house`
- `concrete` under `highrise-house`

looks awesome, but upgrading roads is still TODO...
```
    local tilesMap = {
        ["simple"] = "landfill",
        ["residential"] = "stone-brick",
        ["highrise"] = "concrete",
    }
-- ...
printTiles({...
}, tilesMap[entityName])
```

**Comment by bahrmichael** [comment]
> hmm, dry-dirt for roads looks way better!

`dry-dirt` unfortunately doesn't visually look the same size as the other tiles. I'm not sure if I want to go with dry-dirt but then have to adjust the underlying mapping code so that it looks the same width.

> edit: damn! tile is stone-path, not stone-brick

Thank you for the pointer! I would have stumbled into that one too :D

**Comment by bahrmichael** [comment]
I've got a prototype that upgrades the 5x5 cells around the new residential house. It works, but it doesn't look organic yet. E.g. a circle of cells may be a better fit, and the upgrade should go outward from that circle's center.

Screenshot 2024-03-30 at 21 43 35

**Comment by winex** [comment]
> > hmm, dry-dirt for roads looks way better!
> 
> `dry-dirt` unfortunately doesn't visually look the same size as the other tiles. I'm not sure if I want to go with dry-dirt but then have to adjust the underlying mapping code so that it looks the same width.

oh, i see now - i always play with `show-raw-tile-transitions` (in F4) enabled, so didn't see that `dry-dirt` is narrower, though it is only visually:
scrot 20240331_021605 dry-dirt x2
it still looks better in map view with `landfill` being under buildings

_why on this screenshot newly-built treasury has `tile-unknown` below?! i'm working on consumption currently :)..._ hah, we just need to add it to tile mapping table, because it's a building, too!

**Comment by winex** [comment]
this prototype is interesting. though, i thought house upgade code call `clearCell()` and `printTiles()` with new tiles, didn't it?

File: city.lua [hunk 1]
```diff
@@ -1092,6 +1074,25 @@ local function growCitizenCount(city, count, tier)
...
+    -- "garden"
+    if buildingType == "simple" then
+        return Constants.GROUND_TILE_TYPES.simple
+    elseif buildingType == "residential" then
+        return Constants.GROUND_TILE_TYPES.residential
+    elseif buildingType == "highrise" then
+        return Constants.GROUND_TILE_TYPES.highrise
+    else
+        return Constants.GROUND_TILE_TYPES.simple
+    end
```
Comment (winex) [inline city.lua:1093]: i don't get why you use these ugly string comparisons here again after you've added neat `GROUND_TILE_TYPES` table in `constants.lua` (below)

probably the only good thing about Lua is that it allows named field access both ways like this:
```
-- these are identical:
t = { a = 1, b = 2 }
t = { ["a"] = 1, ["b"] = 2 }
-- access/assignment is identical, too:
t.a, t.b
t["a"], t["b"]
```
don't remember where i've read that, though. that's language syntax, probably

so, this function actually becomes a one-liner:
```
return Constants.GROUND_TILE_TYPES[buildingType] or Constants.GROUND_TILE_TYPES.simple
```

Comment (bahrmichael) [reply city.lua:1093]: Yeah, that looks better! I'll change it.

**Comment by bahrmichael** [comment]
https://github.com/bahrmichael/factorio-tycoon/pull/299/commits/1e5fd2c86741730f9f7479ee91b1d09e470d2d14 now leads to the picture below (with max supply for all tiers). I felt the growth was much smoother.

Screenshot 2024-03-31 at 12 38 31

**Comment by bahrmichael** [comment]
> stone-brick and concrete consumption (and payment) would also be nice for such upgrading, but it's another part -- let me work on this after my consumption PR (it's coming, currently rebasing on your perfomance update)

We can create an issue to keep track of the work, and make me remember that you wanted to work on that if you like :) I do however have some strong opinions on making sure that players are aware of the resources needed. Did a lot of UI rework recently to make sure that works well for the houses.

File: util.lua [hunk 2]
```diff
@@ -209,6 +209,25 @@ local function findCityByEntityUnitNumber(unitNumber)
...
+            local char = string.sub(value, i, i)
+            if char == "1" then
+                game.surfaces[surface_index or Constants.STARTING_SURFACE_ID].set_tiles({ { name = tileName, position = { x, y } } })
+            end
+            x = x + 1
+        end
+        x = start.x
+        y = y + 1
+    end
+
```
Comment (winex) [inline util.lua:230]: ouch! this is missing function closing `end`, unfortunately

Comment (bahrmichael) [reply util.lua:230]: Oops, fixed that in main. Thanks for the pointer!

**Comment by winex** [comment]
damn! i finally found `dry-dirt` on map...
scrot 20240401_235035 dry-dirt crop half
adding dedicated tile seems not a good idea and tiles cannot be `tint`ed, so we probably leave it as is for some time...

**Pull request merged**